dali/pipeline/executor/executor2/exec2.h

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2023-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_EXECUTOR_EXECUTOR2_EXEC2_H_
+#define DALI_PIPELINE_EXECUTOR_EXECUTOR2_EXEC2_H_
+
+#include <memory>
+#include <string>
+#include <vector>
+#include "dali/pipeline/graph/op_graph2.h"
+#include "dali/pipeline/workspace/workspace.h"
+#include "dali/pipeline/executor/executor.h"
+
+namespace dali {
+namespace exec2 {
+
+enum class QueueDepthPolicy : int {
+  FullyBuffered,  //< All operators maintain a queue
+  BackendChange,  //< Only operators followed by one with a different backend have a queue
+  OutputOnly,     //< Only the pipeline output has multiple buffers
+  Legacy = BackendChange,
+};
+
+enum class OperatorConcurrency : int {
+  None,      //< at no time can mutliple operators run
+  Backend,   //< operators from different backends can execute in parallel
+  Full,      //< independent operators can run in parallel
+};
```

Comment:
Is there a concurrency limit on operators?

Reply:
The number of operators that can be run in parallel is limited by the number of threads in the executor.
Otherwise the concurrency is limited by the policy (as described in this enum), the graph topology and the operators being implicitly non-reentrant (you can't run the next iteration of an operator before the previous one is finished).